xls/modules/rle/rle.x

```diff
@@ -0,0 +1,449 @@
+
+    init {(
+        RunLengthEncoderState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev_symbol: bits[SYMBOL_WIDTH]:0,
+            prev_count: bits[COUNT_WIDTH]:0,
+            prev_last: false,
+        }
+    )}
+
+    config (
+        input_r: chan<EncInData<SYMBOL_WIDTH>> in,
+        output_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RunLengthEncoderState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = EncInData {
+          symbol: bits[SYMBOL_WIDTH]:0,
+          last: false
+        };
+        let (input_tok, input) = recv_if(
+            tok, input_r, !state.prev_last, zero_input);
+
+        let prev_symbol_valid = state.prev_count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (
+            input.symbol != state.prev_symbol);
+        let overflow =
+            state.prev_count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let (symbol, count, last) = if (state.prev_last) {
+            (
+                bits[SYMBOL_WIDTH]:0,
+                bits[COUNT_WIDTH]:0,
+                false
+            )
+        } else if (symbol_differ || overflow) {
+            (
+                input.symbol,
+                bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        } else {
+            (
+                input.symbol,
+                state.prev_count + bits[COUNT_WIDTH]:1,
+                input.last,
+            )
+        };
+
+        let data = EncOutData {
+            symbol: state.prev_symbol,
+            count: state.prev_count,
+            last: state.prev_last
+        };
+
+        let do_send = state.prev_last || symbol_differ || overflow;
+        let data_tok = send_if(input_tok, output_s, do_send, data);
+
+        RunLengthEncoderState {
+            prev_symbol: symbol,
+            prev_count: count,
+            prev_last: last,
+        }
+    }
+}
+
+// RLE encoder specialization for the codegen
+proc RLEEnc32 {
+
+    init {()}
+
+    config (
+        input_r: chan<EncInData<32>> in,
+        output_s: chan<EncOutData<32, 2>> out,
+    ) {
+        spawn RLEEnc<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_COMMON_SYMBOL_WIDTH = u32:32;
+// Make counter large enough so that it overflows only in overflow testcase.
+const TEST_COMMON_COUNT_WIDTH = u32:32;
+
+type TestCommonSymbol     = bits[TEST_COMMON_SYMBOL_WIDTH];
+type TestCommonCount      = bits[TEST_COMMON_COUNT_WIDTH];
+type TestCommonEncInData  = EncInData<TEST_COMMON_SYMBOL_WIDTH>;
+type TestCommonEncOutData =
+    EncOutData<TEST_COMMON_SYMBOL_WIDTH, TEST_COMMON_COUNT_WIDTH>;
+
+// Simple transaction without overflow
+const CountSymbolTestSymbolWidth = TEST_COMMON_SYMBOL_WIDTH;
+const CountSymbolTestCountWidth  = TEST_COMMON_COUNT_WIDTH;
+
+type CountSymbolTestStimulus   = TestCommonSymbol;
```

Comment:
curious why we need to prefix the type with the test name since they are basically shared across all tests?

Reply:
it's not necessary for test to work.
For me it was more descriptive to have testcase specific types.

Reply:
@proppy wdyt?

Reply:
That makes sense, thanks for providing the rational.